src/Compilers/CSharp/Portable/Parser/Lexer.cs

```diff
@@ -921,11 +921,65 @@ private bool ScanInteger()
             return start < TextWindow.Position;
         }
 
+        // Allows underscores in integers, except at beginning and end
+        private void ScanNumericLiteralSingleInteger(StringBuilder builder, ref bool underscoreInWrongPlace, ref bool usedUnderscore, bool isHex, bool isBinary)
+        {
+            if (TextWindow.PeekChar() == '_')
+            {
+                underscoreInWrongPlace = true;
+            }
+
+            char ch;
+            var lastCharWasUnderscore = false;
+            while (true)
```

Comment:
I just looked into it (instead of blindly stating "it works"), TextWindow.PeekChar returns InvalidCharacter (char.MaxValue) when past the end of input, which SyntaxFacts.Is*Digit returns false for. So both test and mental tracing logic say it should work.